src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,929 @@
+    assert(zero_violation == U(0), "");
+    return lo;
+  }
+
+  /*
+  1. Intuition:
+  Call r the lowest value not smaller than lo that satisfies bits, consider the
+  first bit in r that is different from the corresponding bit in lo:
+    - Since r is larger than lo the bit must be 0 in lo and 1 in r
+    - Since r must satisify bits the bit must be 0 in zeros
+    - Since r should be the smallest value, this bit should be the lowest one
+      possible
+
+  E.g:      1 2 3 4 5 6
+       lo = 1 0 0 1 1 0
+        x = 1 0 1 0 1 0
+        y = 0 1 1 1 1 1
+  x would be larger than lo since the first different bit is the 3rd one,
+  while y is smaller than lo because the first different bit is the 1st bit.
+  Next, consider:
+       x1 = 1 0 1 0 1 0
+       x2 = 1 0 0 1 1 1
+  Both x1 and x2 are larger than lo, but x1 > x2 since its first different
+  bit from lo is the 3rd one, while with x2 it is the 7th one. As a result,
+  if both x1 and x2 satisfy bits, x2 would be closer to our true result.
+
+  2. Formality:
+  Call i the largest bit index such that:
+
+  - lo[x] satisfies bits for 0 <= x < i (2.1)
+  - zeros[i] = 0                        (2.2)
+  - lo[i] = 0                           (2.3)
+
+  Consider v:
+
+  - v[x] = lo[x], for 0 <= x < i        (2.4)
+  - v[i] = 1                            (2.5)
+  - v[x] = ones[x], for x > i           (2.6)
+
+  We will prove that v is the smallest value not smaller than lo that
+  satisfies bits.
+
+  Call r the smallest value not smaller than lo that satisfies bits. Since lo
+  does not satisfy bits, lo < r (2.7)
+
+  a. Firstly, we prove that r <= v:
+
+    Trivially, lo < v since:
+      lo[x] = v[x], for 0 <= x < i (according to 2.4)
+      lo[i] < v[i] (according to 2.3 and 2.5, lo[i] == 0 < v[i] == 1)
+      bits at x > i have lower significance, and are thus irrelevant
+
+    As established above, the first (i + 1) bits of v satisfy bits.
+    The remaining bits satisfy zeros, since any bit x > i such that zeros[x] == 1,
+      v[x] == ones[x] == 0 (according to 2.6, we assume bits is not contradictory here)
+    They also satisfy ones, since any bit j > i such that ones[x] == 1, v[x] == ones[x] == 1 (according to 2.6)
+
+    As a result, v > lo and v satisfies bits since all of its bits satisfy bits. Which
+    means r <= v since r is the smallest such value.
+
+  b. Secondly, we prove that r == v. Suppose r < v:
+
+    Since r < v, there must be a bit position j that:
+
+    r[j] == 0               (2.b.1)
+    v[j] == 1               (2.b.2)
+    r[x] == v[x], for x < j (2.b.3)
+
+    - If j < i
+      r[j] == 0 (according to 2.b.1)
+      v[j] == lo[j] (according to 2.4 because j < i)
+      v[j] == 1 (according to 2.b.2)
+      r[x] == v[x], for x < j (according to 2.b.3)
+      v[x] == lo[x], for x < j (according to 2,4 because j < i)
+
+      This means that:
+      r[j] == 0
+      lo[j] == 1
+      r[x] == lo[x], for x < j
```

Comment:
```suggestion
      r[j] == 0    (according to 2.b.1)
      lo[j] == 1   (lo[j] == v[j] == 1, according to 2.4 because j < i, and according to 2.b.2)
      r[x] == lo[x], for x < j  (according to 2.4 because j < i)
```
Then you could even drop the other list above, right?